README.md
Early Dynamics of Chronic Myeloid Leukemia on Nilotinib Predicts Deep Molecular Response.
====
This readme text provides you how you run the application of the early prediction of tyrosine kinase inhibitor's response using time-series data of white blood cells (WBCs) and IS values. The enclosed codes were used in [our paper](). (The link is under preparation).
The following text provides you the details of enclosed files and codes; the outputs of our application; requirements of computational environment; how you install and run our codes.

## Description
- ``data/`` - A directory which includes input data, i.e, the time-series of IS and WBC. At the time you get this directory,  all data within this directory are sample data which was generated artificially by ``mkValidationData.py``. When you apply our codes to the real clinical dataset, replace the sample with the data you use.

- ``data/IS_data.csv`` - Time-series data of IS values. Each row corresponds to each patient. Each cell of a row shows the IS value \[\%\] as of 0, 3, ...,24 months of the patient. The left end shows the patient ID and the top end shows the months.

- ``data/WBC_data.csv`` - Time-series data of WBC values. Each row corresponds to each patient. Each cell of a row shows the WBC counts \[num /micro l\] as of 0, 3, ...,24 months of the patient. The left end shows the patient ID and the top end shows the months.

- ``intermediate_data/`` - A directory which contains files corresponding to the optimized parameters and the results of the cross-validation test. These contained files are updated when you run some codes.

- ``out/`` - A directory which contains figure files. They are the time series of IS and WBCs, the scatter diagram of the optimized parameters, and bar graphs showing accuracy, sensitivity, and specificity for several prediction methods with the cross-validation test. These contained figures are updated when you run some codes.

- ``sample_img/`` - Samples of the above output figures.

- ``optimize_all.py`` - This code is used to estimate the parameters *a*,*b*,*c*,*xs*, and *ys* by using all the patients data.
These parameters are saved in the file ``intermediate_data/learned_data.csv``.
We note you have to prepare the files ``data/IS_data.csv`` and ``data/WBC_data.csv`` before you run this code.

- ``optimize_cross.py`` - This code is used to estimate the parameters *a*,*b*,*c*,*xs*, and *ys* for the leave-one-out cross-validation test. These parameters are saved in the file ``intermediate_data/learned_dataXX.csv``, where XX corresponds to the patient ID excluded in the parameter estimation.
We note you have to prepare the files ``data/IS_data.csv`` and ``data/WBC_data.csv`` before you run this code.

- ``visualization.py`` - This code is used to generate figures. The figure ``out/Estimeted_time_seriesXX.svg`` is the estimated time series of IS and WBCs. A figure ``out/parametersYYY.svg`` is the scatter diagram of the estimated parameters. 
We note you have to prepare the files ``data/IS_data.csv``,  ``data/WBC_data.csv``, and ``intermediate_data/learned_data.csv`` files before you run this code.

- ``validation.py`` - This code is used to generate a figure ``out/Score_YYY.svg`` that is the bar graph showing accuracy, sensitivity, and specificity for several prediction methods with the cross-validation test. We note you have to prepare the files
``data/IS_data.csv``, ``data/WBC_data.csv``, and ``intermediate_data/learned_dataXX.csv`` files before you run this code.

- ``mkValidationData.py`` - This code is used to artificially generate sample data of the IS values and WBC counts. These data are saved in``data/IS_data.csv`` and  ``data/WBC_data.csv``, respectively.
***REMARK*** this code overwrites all the data within the ``data/`` directory.  


---
The following files are generated after you run our codes.

- ``intermediate_data/learned_data.csv`` - Estimated parameters are summarized in this csv file. Each column corresponds to each patient. The calculated parameters *a*,*b*,*c*,*xs*,*ys*, and values of the loss function at the end of the calculation are shown for each column.

- ``intermediate_data/learned_dataXX.csv`` - Estimated parameters are summarized in this csv file in the case of the leave-ont-out cross-validation test. XX corresponds to the patient ID which was excluded in the estimation of optimize_cross.py.
- ``intermediate_data/Score_YYY.csv`` - Classification of the prediction results for several methods in terms of the  leave-one-out cross-validation test.YYY represents the criterion of the effectiveness of drug response for CML patients. Detail definitions are introduced in the method section of our paper.

- ``intermediate_data/ref_para.csv`` - Parameter set used in artificial data generation ``mkValidationData.py``.

- ``out/Estimeted_time_seriesXX.svg`` - Output figures of the estimated and observed time series of IS values and WCSs of the XX patient.
The optimized parameters *a*,*b*,*c*,*xs*, and *ys* are used in the calculation of these estimated time series

- ``out/parametersYYY.svg`` - Output figures of the scatter diagram of the optimized parameters *a*,*b*,*xs*, and *ys*. YYY represents the criterion of the effectiveness of drug response for CML patients. Detail definitions are introduced in the method section of our paper.

- ``out/Score_YYY.svg`` - Output figures of the  bar graphs showing accuracy, sensitivity, and specificity for several prediction methods with the cross-validation test. YYY represents the criterion of the effectiveness of drug response for CML patients. Detail definitions are introduced in the method section of our paper.


## Requirement
The enclosed codes were checked by macOS Catalina 10.15.3.
And they require the following environment:

- python 3.7.4
- pandas 0.25.2              
- numpy 1.17.3              
- matplotlib 3.1.1               
- tensorflow 1.15.0

If you already installed ``pip`` 19.0 or later, please execute the following one-sentence code:
```code
pip install pandas matplotlib tensorflow==1.15
```
This code usually ends within several minutes.

## Install
Download the current directory.

## Usage and Demo
1. Before you apply the above codes to your real data-set, replace ``data/IS_data.csv`` and ``data/WBC_data.csv`` with your data.

2. Run the following codes on the current directory:
``` codes
python optimize_all.py
python visualization.py
```

Then, you can get the estimated time series of IS and WBCs as 

<img src="sample_img/time_series_sample.jpg" width="250px">

``out/Estimeted_time_seriesXX.svg``

and the  scatter diagrams of the estimated parameters as 

<img src="sample_img/parameters_sample.jpg" width="550px">


``out/parametersXXX.svg``

Above files are found in the ``out/`` directory.
These codes usually end within 4-6 hours.


3. Run the following codes:
``` codes
python optimize_cross.py
python validation.py
```
Then, you can get the bar graphs showing accuracy, sensitivity, and specificity for several prediction methods with the cross-validation test as

<img src="sample_img/Score_sample.jpg" width="400px">

``out/Score_XXX.svg``

They are found in the ``out/`` directory.
This code usually ends within 5-6 days.

## License
   Copyright 2020 Yuji Okamoto @ The University of Tokyo

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License. 

## Version
0.1.0

The change log file is [here](CHANGELOG.md).

## Code Author

Yuji Okamoto : yuji.0001@gmail.com

Masashi K. Kajita : masashi.k.kajita@gmail.com


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.


## 0.1.0 - 2020-02-14
 - First version

optimize_all.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import numpy as np
import pandas as pd
# import tensorflow  as tf
import tensorflow.compat.v1 as tf # for TensorFlow 2
tf.disable_v2_behavior()
import time


# setting input directory
DATA_DIR = 'data/'



# open data
IS_data =  pd.read_csv(DATA_DIR + 'IS_data.csv',index_col=0).T
WBC_data =  pd.read_csv(DATA_DIR + 'WBC_data.csv',index_col=0).T

# drop nan data
IS_data = IS_data.dropna(axis=1)
WBC_data = WBC_data.dropna(axis=1)
keys = IS_data.keys() & WBC_data.keys()

IS_data = IS_data[keys]
WBC_data = WBC_data[keys]


# object to math type

IS_data.index =  IS_data.index.map(int)
WBC_data.index =  WBC_data.index.map(int)


# To numpy data
IS_values = IS_data.values
WBC_values = WBC_data.values

# Number of Patients
Np = IS_values.shape[1]


# Definition of MR4.0, DMR, CMR
# 1: Cure, 0: Non cure 

CMR_values =np.array((IS_values[8,:]<=0.0032) & (IS_values[7,:]<=0.0032),dtype = np.int32)
DMR_values =np.array(IS_values[8,:]<=0.0032,dtype = np.int32)
MR4_values =np.array(IS_values[4,:] <= 0.01,dtype = np.int32)


# Optimization 

# Learning parameters:
# scalar value : c
# vector value : a, b, xs, ys
# 
# explanatory value : IS(0M), WBC(0M)
# objective value : \sum_{t = 3M, ... , 24M} log(IS(t))
# estimation func :  mean squre error of  log(IS(t))

# Setting of parameters
xs_min = 3100
xs_max = 9800
epsilon = 0.00000001



# Designing models

# change time scaling from [3 months, 6 months, ...] to [1,2,...]
times = tf.reshape(tf.constant([1,2,3,4,5,6,7,8],tf.float32),[8,1])

# Initialization of parameters
xs_tmp  = tf.Variable(np.ones(Np),dtype=tf.float32)
xs =  6000 * xs_tmp
a = tf.Variable(0.5 * np.ones(Np),dtype=tf.float32)

ys_tmp  = tf.Variable(np.ones(Np),dtype=tf.float32)
ys =  10 * ys_tmp
b  = tf.Variable( 5*np.ones(Np),dtype=tf.float32)

c =  tf.Variable(1.0, dtype=tf.float32)

# Initialization of placeholders
_IS = tf.placeholder(tf.float32, shape = [9,Np])
_WBC = tf.placeholder(tf.float32, shape = [9,Np])

# Calculating x0 and y0 from IS(0) and WBC(0)
x0 =  (120- _IS [0,:]) / (120 + (c - 1) *_IS [0,:])* _WBC[0,:]
y0 =  c* _IS[0,: ] / (120 + (c - 1)*_IS [0,:])* _WBC[0,:]

# Calculating x(t) and y(t) from times, x0 and y0
xt  =  xs +  (x0  -  xs) * tf.exp(-a* times)
yt  =  ys +  (y0  -  ys) * tf.exp(-b* times)

# Observation of IS(t) and WBC(t). 
# 0.0032 is observation limit of IS values.
IS =  tf.maximum(yt / (c * xt + yt) * 120, 0.0032)
# If WBC == 0, np.log(WBC) == inf in loss function,
WBC =  tf.maximum(xt + yt,epsilon)

# ys minimal values estimated by the observation limit of IS.
ys_min = xs *0.0032/120*c

# Constraint conditions
# Condition 1 : upper and lower bound of normal WBCs
sc1 = tf.reduce_sum(tf.nn.relu( - xs + xs_min))  + tf.reduce_sum(tf.nn.relu( xs - xs_max))

# Condition 2 : parameters a and b is positive
sc2 = tf.reduce_sum(tf.nn.relu( - a)) + tf.reduce_sum(tf.nn.relu( - b)) 

# Condition 3 : parameters ys is larger than ys_min
sc3 = tf.reduce_sum(tf.nn.relu( - ys+ys_min))



# Designing loss function
loss  = tf.reduce_mean(tf.square( tf.log(IS) - tf.log(_IS[1:,:]))) + tf.reduce_mean(tf.square( tf.log(WBC) - tf.log(_WBC[1:,:]))) + 100 * (sc1 +  sc2 + sc3)
# Setting optimizer
train =  tf.train.AdamOptimizer(0.00001).minimize(loss)


# Optimization using all partients data

# caluclation starting
start_time = time.time()

# Iteration 
## if you want to finish the optimization quickly, you can choose the lower  iteration number N.
N = 10000000
# N = 10000


# Initialization
sess = tf.Session()
sess.run(tf.global_variables_initializer())

# Setting learning data
IS_learn  = IS_values
WBC_learn = WBC_values
keys_learn = IS_data.keys()
# learning
for  i in range(N):
    sess.run(train,feed_dict={_IS:IS_learn,_WBC:WBC_learn})
    if (i%(N//100) == 0):
        print('{0:02d} %'.format(i//(N//100)))
loss_data = sess.run(loss, feed_dict={_IS:IS_learn,_WBC:WBC_learn})
# get parameters(time re-scaling )
c_learn = sess.run(c) 
a_learn = sess.run(a) / 3
xs_learn = sess.run(xs)
b_learn =  sess.run(b) / 3
ys_learn =  sess.run(ys)

x0_learn =  sess.run(x0, feed_dict={_IS:IS_learn,_WBC:WBC_learn})
y0_learn =  sess.run(y0, feed_dict={_IS:IS_learn,_WBC:WBC_learn})

# save data
learned_values = np.vstack((xs_learn,ys_learn,a_learn,b_learn,c_learn* np.ones(xs_learn.shape),loss_data* np.ones(xs_learn.shape)))
learned_data  =  pd.DataFrame(learned_values,index =['xs','ys','a','b','c','loss_data'], columns=keys_learn )
learned_data.to_csv('intermediate_data/learned_data.csv')


# Calculation finishing
elapsed_time = time.time() - start_time
print('elapsed_time ={0:.04f}'.format(elapsed_time))





optimize_cross.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import numpy as np
import pandas as pd
# import tensorflow  as tf
import tensorflow.compat.v1 as tf # for TensorFlow 2
tf.disable_v2_behavior()
import time

# setting input directory
DATA_DIR = 'data/'

# open data
IS_data =  pd.read_csv(DATA_DIR + 'IS_data.csv',index_col=0).T
WBC_data =  pd.read_csv(DATA_DIR + 'WBC_data.csv',index_col=0).T

# drop nan data
IS_data = IS_data.dropna(axis=1)
WBC_data = WBC_data.dropna(axis=1)
keys = IS_data.keys() & WBC_data.keys()

IS_data = IS_data[keys]
WBC_data = WBC_data[keys]


# object to math type

IS_data.index =  IS_data.index.map(int)
WBC_data.index =  WBC_data.index.map(int)

# To numpy data
time_vales = np.array(IS_data.index,dtype = int)
IS_values = IS_data.values
WBC_values = WBC_data.values

# Number of Patients
Np = IS_values.shape[1]


# Definition of MR4.0, DMR, CMR
# 1: Cure, 0: Non cure 

CMR_values =np.array((IS_values[8,:]<=0.0032) & (IS_values[7,:]<=0.0032),dtype = np.int32)
DMR_values =np.array(IS_values[8,:]<=0.0032,dtype = np.int32)
MR4_values =np.array(IS_values[4,:] <= 0.01,dtype = np.int32)


# Optimization 

# Learning parameters:
# scalar value : c
# vector value : a, b, xs, ys
# 
# explanatory value : IS(0M), WBC(0M)
# objective value : \sum_{t = 3M, ... , 24M} log(IS(t))
# estimation func :  mean squre error of  log(IS(t))

# Setting of parameters
xs_min = 3100
xs_max = 9800
epsilon = 0.00000001


# Designing model

# change time scaling from [3 months, 6 months, ...] to [1,2,...]
times = tf.reshape(tf.constant([1,2,3,4,5,6,7,8],tf.float32),[8,1])

# Initirizaiton of parameters
xs_tmp  = tf.Variable(np.ones(Np-1),dtype=tf.float32)
xs =  6000 * xs_tmp
a = tf.Variable(0.5 * np.ones(Np-1),dtype=tf.float32)

ys_tmp  = tf.Variable(np.ones(Np-1),dtype=tf.float32)
ys =  10 * ys_tmp
b  = tf.Variable( 5*np.ones(Np-1),dtype=tf.float32)

c =  tf.Variable(1.0, dtype=tf.float32)

# Initizization of placeholders
_IS = tf.placeholder(tf.float32, shape = [9,Np-1])
_WBC = tf.placeholder(tf.float32, shape = [9,Np-1])

# Calculating x0 and y0 from IS(0) and WBC(0)
x0 =  (120- _IS [0,:]) / (120 + (c - 1) *_IS [0,:])* _WBC[0,:]
y0 =  c* _IS[0,: ] / (120 + (c - 1)*_IS [0,:])* _WBC[0,:]

# Calculating x(t) and y(t) from times, x0 and y0
xt  =  xs +  (x0  -  xs) * tf.exp(-a* times)
yt  =  ys +  (y0  -  ys) * tf.exp(-b* times)

# Observation of IS(t) and WBC(t). 
# 0.0032 is observation limit of IS values.
IS =  tf.maximum(yt / (c * xt + yt) * 120, 0.0032)
# If WBC == 0, np.log(WBC) == inf in loss function,
WBC =  tf.maximum(xt + yt,epsilon)

# ys minimal values estimated by the observation limit of IS.
ys_min = xs *0.0032/120*c

# Constraint conditions
# Condition 1 : upper and lower bound of normal WBCs
sc1 = tf.reduce_sum(tf.nn.relu( - xs + xs_min))  + tf.reduce_sum(tf.nn.relu( xs - xs_max))

# Condition 2 : parameters a and b is positive
sc2 = tf.reduce_sum(tf.nn.relu( - a)) + tf.reduce_sum(tf.nn.relu( - b)) 

# Condition 3 : parameters ys is larger than ys_min
sc3 = tf.reduce_sum(tf.nn.relu( - ys+ys_min))


# Designing loss function
loss  = tf.reduce_mean(tf.square( tf.log(IS) - tf.log(_IS[1:,:]))) + tf.reduce_mean(tf.square( tf.log(WBC) - tf.log(_WBC[1:,:]))) + 100 * (sc1 +  sc2 + sc3)
# Setting optimizer
train =  tf.train.AdamOptimizer(0.00001).minimize(loss)



# Optimization using all partients data

# caluclation starting
start_time = time.time()

# Iteration 
## if you want to finish the optimization quickly, you can choose the lower  iteration number N.
N = 10000000
# N = 10

learned_data = []
# Initialization
sess = tf.Session()

for i_learn in range(Np):
    #Initialization weights
    sess.run(tf.global_variables_initializer())
    # Setting learning data
    IS_learn  = np.hstack([IS_values[:,:i_learn],IS_values[:,i_learn+1:]])
    WBC_learn = np.hstack([WBC_values[:,:i_learn],WBC_values[:,i_learn+1:]])
    keys_learn = np.hstack([IS_data.keys()[:i_learn],IS_data.keys()[i_learn+1:]])
    # learning
    for  i in range(N):
        sess.run(train,feed_dict={_IS:IS_learn,_WBC:WBC_learn})
    loss_data = sess.run(loss, feed_dict={_IS:IS_learn,_WBC:WBC_learn})    
    # get parameters(time re-scaling )
    c_learn = sess.run(c) 
    a_learn = sess.run(a) / 3
    xs_learn = sess.run(xs)
    b_learn =  sess.run(b) / 3
    ys_learn =  sess.run(ys)

    x0_learn =  sess.run(x0, feed_dict={_IS:IS_learn,_WBC:WBC_learn})
    y0_learn =  sess.run(y0, feed_dict={_IS:IS_learn,_WBC:WBC_learn})


    # save data
    learned_values = np.vstack((xs_learn,ys_learn,a_learn,b_learn,c_learn* np.ones(xs_learn.shape),loss_data* np.ones(xs_learn.shape)))
    learned_data.append(pd.DataFrame(learned_values,index =['xs','ys','a','b','c','loss_data'], columns=keys_learn ))
    learned_data[i_learn].to_csv('intermediate_data/learned_data{0:02d}.csv'.format(i_learn))

    print('{0:02d} th calc is finished.'.format(i_learn))

# Calculation finishing
elapsed_time = time.time() - start_time
print('elapsed_time ={0:.04f}'.format(elapsed_time))




mkValidationData.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import numpy as np
import pandas as pd


# In[2]:


# number of validation data
# N_DMR : Number of DMR patients
# N_NDMR : Number of non DMR patients

N_DMR = 11
N_NDMR = 11
N = N_DMR + N_NDMR
# drop number of IS values
N_IS_drop = 2
# drop number of WBCs
N_WBC_drop = 1

# other setting
c = 1.25
t = np.array([0,3,6,9,12,15,18,21,24])


# In[3]:


# making parameter of validation data
a = np.random.normal(3,1,N)
a[a<1]=1
xs = np.random.normal(6000,1000,N)

b = np.r_[np.random.normal(3,1,N_DMR),np.random.normal(2,1,N_NDMR)]
b[b<0.5]=0.5

ys = np.r_[np.random.normal(-1,0.1,N_DMR),np.random.normal(0.5,0.3,N_NDMR)]
ys = 10**ys

x0 = np.random.normal(3000,500,N)
x0[x0<100]=100
y0 = 10**np.random.normal(5,0.5,N)
y0[y0<10000]=10000




# In[4]:


# Calculating IS values and WBCs
xt = xs +  (x0 - xs)*np.e**(-a*t.reshape(-1,1))
yt = ys +  (y0 - ys)*np.e**(-b*t.reshape(-1,1))
ISt = yt/(c*xt + yt)*1.2*100
ISt[ISt<0.0032] = 0.0032
WBCt = xt + yt


# In[5]:


# numpy array to pandas dataframe 
IS_data = pd.DataFrame(ISt.T,columns=t)
WBC_data = pd.DataFrame(WBCt.T,columns=t)

# Add drop
IS_drop = np.c_[t[np.random.randint(0,len(t),N_IS_drop)],np.random.randint(0,N,N_IS_drop)]
WBC_drop = np.c_[t[np.random.randint(0,len(t),N_WBC_drop)],np.random.randint(0,N,N_WBC_drop)]

for i in range(N_IS_drop):
    IS_data[IS_drop[i][0]][IS_drop[i][1]]=''

for i in range(N_WBC_drop):
    WBC_data[WBC_drop[i][0]][WBC_drop[i][1]]=''


# In[6]:


# saving
IS_data.to_csv('data/IS_data.csv')
WBC_data.to_csv('data/WBC_data.csv')


# In[7]:


# saving reference parameters of validation data
ref_para = pd.DataFrame(np.c_[x0,xs,a,y0,ys,b],columns=["x0","xs","a","y0","ys","b"])
ref_para.to_csv('intermediate_data/ref_para.csv')



visualization.py
#!/usr/bin/env python
# coding: utf-8

# In[9]:


import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import matplotlib.patches as patches


# In[10]:


# setting input directory
DATA_DIR = 'data/'
IDATA_DIR = 'intermediate_data/'
IMG_DIR = 'out/'


# In[11]:


# open data
IS_data =  pd.read_csv(DATA_DIR + 'IS_data.csv',index_col=0).T
WBC_data =  pd.read_csv(DATA_DIR + 'WBC_data.csv',index_col=0).T

# drop nan data
IS_data = IS_data.dropna(axis=1)
WBC_data = WBC_data.dropna(axis=1)
keys = IS_data.keys() & WBC_data.keys()

IS_data = IS_data[keys]
WBC_data = WBC_data[keys]


# object to math type

IS_data.index =  IS_data.index.map(int)
WBC_data.index =  WBC_data.index.map(int)


# To numpy data
time_values = np.array(IS_data.index,dtype = int)
IS_values = IS_data.values
WBC_values = WBC_data.values

# Number of Patients
Np = IS_values.shape[1]


# check response type
dict_MR4 = {0:' Non MR4.0 ',1:'MR4.0'}
dict_MR45 = {0:' Non MR4.5 ',1:'MR4.5'}
dict_CMR = {0:' Non CMR ',1:'CMR'}

dict_list = [dict_MR4,dict_MR45,dict_CMR]

MR4_values = np.array(np.sum((IS_values<=0.01),axis=0)>0,dtype=np.int32) # MR4.0
MR45_values = np.array(np.sum((IS_values<=0.0032),axis=0)>0,dtype=np.int32)# MR4.5
# CMR_values =np.array((IS_values[8,:]==0.0032) & (IS_values[7,:]==0.0032),dtype = np.int32)
CMR_values =np.array((IS_values[8,:]<=0.0032) & (IS_values[7,:]<=0.0032),dtype = np.int32)

MR_list = [MR4_values,MR45_values,CMR_values]


# In[12]:



# import learned data
df_dynamics_parameter  =  pd.read_csv(IDATA_DIR + 'learned_data.csv',index_col=0)


# estimate time-series
epsilon = 0.00000001

c_learn =  df_dynamics_parameter.loc['c'].values[0]
xs_learn = df_dynamics_parameter.loc['xs'].values
x0_learn = (120 - IS_values[0,:])/(120 +(c_learn-1)*IS_values[0,:]) * WBC_values[0,:]

ys_learn = df_dynamics_parameter.loc['ys'].values
y0_learn =  c_learn * IS_values[0,:]/(120 +(c_learn-1)*IS_values[0,:]) * WBC_values[0,:]
 
a_learn  = df_dynamics_parameter.loc['a'].values
b_learn  = df_dynamics_parameter.loc['b'].values

xt_est = xs_learn + (x0_learn - xs_learn) * np.exp(- a_learn * np.reshape(time_values,[len(time_values),1]))
yt_est = ys_learn + (y0_learn - ys_learn) * np.exp(- b_learn *np.reshape(time_values,[len(time_values),1]))

 
IS_est = np.maximum(120 * yt_est/np.maximum(c_learn * xt_est + yt_est,epsilon),0.0032)
WBC_est = xt_est + yt_est


# In[13]:



# plot estimated time-series of IS and WBC

fig_list = IS_data.keys()

for tmp in fig_list:
    i =  np.where(IS_data.keys() == tmp)[0][0]
    plt.close()
    fig = plt.figure(figsize=(1.6, 1.6)).patch.set_alpha(0)

    plt.subplot(2,1,1)
    plt.plot(time_values,IS_values[:,i],'-s',color ='black',label = 'Measured data',linewidth = 1,markersize = 3)
    plt.plot(time_values,IS_est[:,i],':^',color ='red', label = 'Estimated data',linewidth = 1,markersize = 3)

    plt.xlim([0,24])
    plt.ylim([0.002,150])

    plt.tick_params(labelsize=6)
    plt.gca().yaxis.set_label_coords(-0.19, 0.5)
    plt.ylabel('IS [%]',fontsize=6)

    plt.title('Pt. {0:02d}'.format(WBC_data.keys()[i]),fontsize = 6)
    plt.yscale('log')


    plt.gca().spines['right'].set_visible(False)
    plt.gca().spines['top'].set_visible(False)
    plt.gca().yaxis.set_ticks_position('left')
    plt.gca().xaxis.set_ticks_position('bottom')
    plt.xticks([])
    plt.yticks([0.1,10])
    plt.gca().set_yticklabels(['0.1','10'])
    plt.legend(fontsize=6)



    plt.subplot(2,1,2)
    plt.plot(time_values,WBC_values[:,i],'-s',color ='black',linewidth = 1,markersize = 3)
    plt.plot(time_values,WBC_est[:,i],':^',color ='red',linewidth = 1,markersize = 3)
    plt.gca().add_patch(patches.Rectangle(xy = (0, 3100), width = 24, height = 6700,angle = 0, color = "cyan", alpha = 0.3,label = 'Normal level'))
    plt.xlim([0,24])
    plt.xticks([0,3,6,9,12,15,18,21,24])
    plt.ylim([2000,100000])
    plt.gca().yaxis.set_label_coords(-0.19, 0.5)
    plt.ylabel('WBCs [num/$\mathrm{\mu}$l]',fontsize = 6)
    plt.gca().xaxis.set_label_coords(0.5, -0.4)
    plt.xlabel('Time [month]',fontsize=6)
    plt.yscale('log')

    plt.gca().spines['right'].set_visible(False)
    plt.gca().spines['top'].set_visible(False)
    plt.gca().yaxis.set_ticks_position('left')
    plt.gca().xaxis.set_ticks_position('bottom')
    plt.tick_params(labelsize=6)
    plt.legend(fontsize=6)

    plt.savefig(IMG_DIR +  'Estimeted_time_series'+'{0:02d}'.format(tmp) + '.svg',bbox_inches='tight', pad_inches=0)



# In[14]:


# function making xs-a plot 
def mkxFig(DMR_values,dict_DMR):
    plt.scatter(xs_learn[DMR_values==0], a_learn[DMR_values==0],s = 20,color = 'red',label = dict_DMR[0][1:])
    plt.scatter(xs_learn[DMR_values==1], a_learn[DMR_values==1],s = 20,color = 'black',label =  dict_DMR[1])

    for i in range(Np):
        plt.text(xs_learn[i], a_learn[i]+0.1, WBC_data.keys()[i],fontsize = 6)


    plt.xlim([3000,10000])
    plt.xticks([4000,6000,8000,10000])
    plt.tick_params(labelsize=6)

    plt.gca().spines['right'].set_visible(False)
    plt.gca().spines['top'].set_visible(False)
    plt.gca().yaxis.set_ticks_position('left')
    plt.gca().xaxis.set_ticks_position('bottom')
    plt.legend(fontsize=6,bbox_to_anchor=(0.6, 0.2), loc='upper right', borderaxespad=0)


    plt.gca().xaxis.set_label_coords(0.5, -0.15)    
    plt.xlabel('Convergence value of \n normal WBCs [num/$\mathrm{\mu}$l]',fontsize = 6)
    plt.ylabel('Recovery rate of  \n normal WBCs [1/month]',fontsize = 6)






# In[15]:


# function making ys-b plot 
def mkyFig(DMR_values,dict_DMR):
    plt.scatter(ys_learn[DMR_values==0], b_learn[DMR_values==0],s = 20,color = 'red',label = dict_DMR[0][1:])
    plt.scatter(ys_learn[DMR_values==1], b_learn[DMR_values==1],s = 20,color = 'black',label =  dict_DMR[1])

    for i in range(Np):
        plt.text(ys_learn[i], b_learn[i]+0.1, WBC_data.keys()[i],fontsize = 6)

    plt.xscale('log')
    # plt.xlim([0.1,300])
    plt.xlim([0.05,300])
    # plt.ylim([0,7.5])
    plt.ylim([0,9.0])
    plt.legend()
    plt.gca().xaxis.set_label_coords(0.5, -0.15)
    plt.xlabel('Convergence value of \n CML cells [num/$\mathrm{\mu}$l]',fontsize = 6)
    plt.ylabel('Reduction rate of \n CML cells [1/month]',fontsize = 6)

    plt.gca().spines['right'].set_visible(False)
    plt.gca().spines['top'].set_visible(False)
    plt.gca().yaxis.set_ticks_position('left')
    plt.gca().xaxis.set_ticks_position('bottom')
    plt.tick_params(labelsize=6)
    plt.legend(fontsize=6)





# In[16]:



for i in range(3):

    plt.close()
    plt.figure(figsize=(5.0, 2.1)).patch.set_alpha(0)
    plt.subplots_adjust(wspace=0.4, hspace=0)
    plt.subplot(1,2,1)
    mkxFig(MR_list[i],dict_list[i])
    plt.subplot(1,2,2)
    mkyFig(MR_list[i],dict_list[i])

    plt.savefig(IMG_DIR + 'parameters' + dict_list[i][1]+'.svg',bbox_inches='tight', pad_inches=0)





